validphys2/src/validphys/photon/compute.py

```diff
@@ -0,0 +1,321 @@
+            for photon_array in self.photons_array
+        ]
+
+    def compute_photon_array(self, replica):
+        r"""
+        Compute the photon PDF for every point in the grid xgrid.
+
+        Parameters
+        ----------
+        replica: int
+            replica id
+
+        Returns
+        -------
+        compute_photon_array: numpy.array
+            photon PDF at the fitting scale Q0
+        """
+        # Compute photon PDF
+        log.info(f"Computing photon")
+        start_time = time.perf_counter()
+        photon_qin = np.array(
+            [self.lux[replica].EvaluatePhoton(x, Q_IN**2).total for x in self.xgrid]
+        )
+        log.info(f"Computation time: {time.perf_counter() - start_time}")
+        photon_qin += self.generate_errors(replica)
+        photon_qin /= self.xgrid
+        # TODO : the different x points could be even computed in parallel
+
+        # Load eko and reshape it
+        with EKO.read(self.path_to_eko_photon) as eko:
+            # TODO : if the eko has not the correct grid we have to reshape it
+            # it has to be done inside vp-setupfit
+
+            # construct PDFs
+            pdfs_init = np.zeros((len(eko.rotations.inputpids), len(self.xgrid)))
+            for j, pid in enumerate(eko.rotations.inputpids):
+                if pid == 22:
+                    pdfs_init[j] = photon_qin
+                    ph_id = j
+                if pid not in self.qcd_pdfs.flavors:
+                    continue
+                pdfs_init[j] = np.array(
+                    [self.qcd_pdfs.xfxQ(x, Q_IN, replica, pid) / x for x in self.xgrid]
+                )
+
+            # Apply EKO to PDFs
+            q2 = eko.mu2grid[0]
+            with eko.operator(q2) as elem:
+                pdfs_final = np.einsum("ajbk,bk", elem.operator, pdfs_init)
+
+        photon_Q0 = pdfs_final[ph_id]
+
+        # we want x * gamma(x)
+        return self.xgrid * photon_Q0
+
+    def __call__(self, xgrid):
+        """
+        Compute the photon interpolating the values of self.photon_array.
+
+        Parameters
+        ----------
+        xgrid : nd.array
+            array of x values with shape (1,xgrid,1)
+
+        Returns
+        -------
+        photon values : nd.array
+            array of photon values with shape (1,xgrid,1)
+        """
+        return [
+            self.interpolator[id](xgrid[0, :, 0])[np.newaxis, :, np.newaxis]
+            for id in range(len(self.replicas_id))
+        ]
+
+    def integrate(self):
+        """Compute the integral of the photon on the x range."""
+        return [
+            trapezoid(self.photons_array[id], self.xgrid)
+            for id in range(len(self.replicas_id))
+        ]
+
+    def generate_error_matrix(self):
+        """generate error matrix to be used for the additional errors."""
+        if not self.fiatlux_runcard["additional_errors"]:
+            return None
+        extra_set = LHAPDFSet(EXTRA_SET, "replicas")
```

Comment:
Actually, now that I think about it. We should `parse_` any pdfset input first such that vp can raise a ConfigError or  download the resource if needed (or not, depending on if the `--no-net` flag is active). So I'm afraid this will have to be changed, same for when loading the pdf from the runcard.

Reply:
FWIW this was main point long ago: That the pdf set should be found (and downloaded) using the config machinery. Not so much about whether you then use our compatibilty wrapper or the lhapdf bindings directly.

Reply:
I think this has already been implemented by Niccolo last week